group-coordinator/src/test/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntimeTest.java

```diff
@@ -0,0 +1,803 @@
+                .build();
+
+        // Loads the coordinator.
+        runtime.scheduleLoadOperation(TP, 10);
+
+        // Verify the initial state.
+        CoordinatorRuntime<MockCoordinator, String>.CoordinatorContext ctx = runtime.contextOrThrow(TP);
+        assertEquals(0L, ctx.lastWrittenOffset);
+        assertEquals(0L, ctx.lastCommittedOffset);
+        assertEquals(Collections.singletonList(0L), ctx.snapshotRegistry.epochsList());
+
+        // Override the coordinator with a coordinator that throws
+        // an exception when replay is called.
+        ctx.coordinator = new MockCoordinator(ctx.snapshotRegistry) {
+            @Override
+            public void replay(String record) throws RuntimeException {
+                throw new IllegalArgumentException("error");
+            }
+        };
+
+        // Write. It should fail.
+        CompletableFuture<String> write = runtime.scheduleWriteOperation("write", TP,
+            state -> new CoordinatorResult<>(Arrays.asList("record1", "record2"), "response1"));
+        assertFutureThrows(write, IllegalArgumentException.class);
+
+        // Verify that the state has not changed.
+        assertEquals(0L, ctx.lastWrittenOffset);
+        assertEquals(0L, ctx.lastCommittedOffset);
+        assertEquals(Collections.singletonList(0L), ctx.snapshotRegistry.epochsList());
+    }
+
+    @Test
+    public void testScheduleWriteOpWhenWriteFails() {
+        // The partition writer only accept on write.
+        MockPartitionWriter writer = new MockPartitionWriter(1);
+
+        CoordinatorRuntime<MockCoordinator, String> runtime =
+            new CoordinatorRuntime.Builder<MockCoordinator, String>()
+                .withLoader(new MockCoordinatorLoader())
+                .withEventProcessor(new MockEventProcessor())
+                .withPartitionWriter(writer)
+                .withCoordinatorBuilderSupplier(new MockCoordinatorBuilderSupplier())
+                .build();
+
+        // Loads the coordinator.
+        runtime.scheduleLoadOperation(TP, 10);
+
+        // Verify the initial state.
+        CoordinatorRuntime<MockCoordinator, String>.CoordinatorContext ctx = runtime.contextOrThrow(TP);
+        assertEquals(0, ctx.lastWrittenOffset);
+        assertEquals(0, ctx.lastCommittedOffset);
+        assertEquals(Collections.singletonList(0L), ctx.snapshotRegistry.epochsList());
+
+        // Write #1. It should succeed and be applied to the coordinator.
+        runtime.scheduleWriteOperation("write#1", TP,
+            state -> new CoordinatorResult<>(Arrays.asList("record1", "record2"), "response1"));
+
+        // Verify that the state has been updated.
+        assertEquals(2L, ctx.lastWrittenOffset);
+        assertEquals(0L, ctx.lastCommittedOffset);
+        assertEquals(Arrays.asList(0L, 2L), ctx.snapshotRegistry.epochsList());
+        assertEquals(mkSet("record1", "record2"), ctx.coordinator.records());
+
+        // Write #2. It should fail.
```

Comment:
This fails because we only allow one write? (That took me a moment to realize)

If we returned 3 records in write2, I think this test would be stronger.

Reply:
The number of records does not matter here. The writer is configured to only accept one successful call to PartitionWriter#append. Let me extend the comment to make this clear.

Reply:
Yeah -- comments would help. I think the other part I was suggesting is that write #2 seems to be the same as write #1. So if it didn't fail due to the write limit, would we have in the records ["record1", "record2", "record1", "record2"]? If that's the case this is fine. 

I originally thought that if the records were the same, it was a no op, but I was confusing things.

Reply:
Yeah, I agree that the records are not great in this test. For the context, the runtime always write what it gets. It does not compare the records themselves.

I have updated the writer to fail if the number of records in a single write is greater than a threshold. I think that it will be less confusing this way. What do you think?